### PR TITLE
[5.4] Adds abstract type method to container

### PR DESCRIPTION
**Pull request by ghost**
Enables users of containers to retrieve all bound types that are not aliases. As alternative to #17618

### CODE REVIEW

File: src/Illuminate/Container/Container.php [hunk 1]
```diff
@@ -940,6 +940,16 @@ public function getBindings()
     }
 
     /**
+     * Return the types which are bound in this container, except for aliases.
+     *
+     * @return array
+     */
+    public function getAbstractTypes()
+    {
+        return array_unique(array_merge(array_keys($this->bindings), array_keys($this->instances)));
```
Comment (CristianLlanos) [inline src/Illuminate/Container/Container.php]: I believe `array_unique` is not needed if you do it like this:
```php
return array_keys(array_merge($this->bindings, $this->instances));
````

Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]: It should also be noted that array_unique has quartic time complexity, so should be avoided.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Container/Container.php (hunk 1 above)
Comment (praalhans) [inline src/Illuminate/Container/Container.php]: Fixed.

W.r.g.t. quartic time complexity, PHP docs state:

> array_unique() sorts the values treated as string at first, then will keep the first key encountered for every value, and ignore all following keys. It does not mean that the key of the first related value from the unsorted array will be kept.

That is worst-case quadratic only if all strings have the same length as the array itself. Could you point me to a reference explicating why it is quartic?

**Comment by taylorotwell** [comment]
The method name doesn't make sense.

**Comment by taylorotwell** [comment]
No plans to add this at least under this method name for now.

**Pull request closed without merging**